src/book.py

```diff
@@ -0,0 +1,224 @@
+from nltk.corpus import stopwords
+from nltk import FreqDist, word_tokenize, sent_tokenize, WordNetLemmatizer
+import re
+import random
+
+
+class BookText():
+    """A class for reading and manipulating texts"""
+
+    def __init__(self, filepath=None, rawtext=None, encoding='utf-8',
+                 clean=False, author=None, title=None, meta=None):
+        """Constructor for BookText
+
+        parameters:
+            **NB** one of filepath or rawtext must be specified
+
+            filepath (None): filepath to a text file to read
+            rawtext (None): a raw text string
+            encoding ("utf-8"): the encoding to use
+            clean (False): whether to clean the text on initializing
+            author (None): if not specified, inferred from text
+            title (None): if not specified, inferred from text
+        """
+        if not (filepath or rawtext):
+            raise ValueError('Must specify one of filepath or rawtext')
+        if filepath is not None:
+            # Reading in the file
+            with open(filepath, encoding=encoding) as f:
+                data = f.read()
+        else:
+            data = rawtext
+        # Using regex to mark the start and end of the book
+        rex_start = r"START OF (THIS|THE) PROJECT GUTENBERG EBOOK (.*) \*\*\*"
+        rex_end = r"(?i)END of (this|the|) Project Gutenberg"
+        try:
+            start_pos = re.search(rex_start, data).span()[1]
+        except AttributeError:  # re.search returned None
+            start_pos = 0
+        try:
+            end_pos = re.search(rex_end, data).span()[0]
+        except AttributeError:  # re.search returned None
+            end_pos = None
+        meta_data = data[:start_pos]
+        text_of_book = data[start_pos:end_pos]
+
+        self._text = text_of_book
+        if clean:
+            self.clean(inplace=True)
+        self.start_pos = start_pos
+        self.end_pos = end_pos
+        self._meta = meta or meta_data
+        if author is None:
+            try:
+                self._author = re.search(
+                    r"(?<=Author: )[\w\s]+(?=\n)", meta_data).group().strip()
+            except AttributeError:
+                self._author = None
+        else:
+            self._author = author
+        if title is None:
+            try:
+                self._title = re.search(
+                    r"(?<=Title: )[\w\s]+(?=\n)", meta_data).group().strip()
+            except AttributeError:
```

Comment:
This seems like something for another time. It also assumes our filenames will always follow our convention, which they may not if we start scraping from other sources, or anonymous sources.

Reply:
This class as been fine tuned to work on Gutenberg Project files which should hopefully all follow similar schema. If we find a file for which it does not work, we should figure out why that file failed and adapt from there.
But, I do agree that maybe if a user inputs an author or title at instantiation, then those should take precedent.

Reply:
I'm confused by your last sentence. If a user inputs author or title, only the input is used

Reply:
The scraping code that I have creates a name for each book based on the author's last name and the title. I don't think that is going to change irrespective of the source we scrape from. Kyle, it might be a good idea to add a few lines to get the title from the filename too, if it's not available from other sources for some reason.

Reply:
Comment addressed in commit b2e4634